src/stream/src/executor/approx_percentile/global_state.rs

```diff
@@ -0,0 +1,340 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::{BTreeMap, Bound};
+use std::mem;
+
+use risingwave_common::array::Op;
+use risingwave_common::bail;
+use risingwave_common::row::Row;
+use risingwave_common::types::{Datum, ToOwnedDatum};
+use risingwave_common::util::epoch::EpochPair;
+use risingwave_storage::store::PrefetchOptions;
+use risingwave_storage::StateStore;
+
+use crate::executor::prelude::*;
+use crate::executor::StreamExecutorResult;
+
+/// The global approx percentile state.
+pub struct GlobalApproxPercentileState<S: StateStore> {
+    quantile: f64,
+    base: f64,
+    row_count: i64,
+    bucket_state_table: StateTable<S>,
+    count_state_table: StateTable<S>,
+    cache: BucketTableCache,
+    last_output: Option<Datum>,
+    output_changed: bool,
+}
+
+// Initialization
+impl<S: StateStore> GlobalApproxPercentileState<S> {
+    pub fn new(
+        quantile: f64,
+        base: f64,
+        bucket_state_table: StateTable<S>,
+        count_state_table: StateTable<S>,
+    ) -> Self {
+        Self {
+            quantile,
+            base,
+            row_count: 0,
+            bucket_state_table,
+            count_state_table,
+            cache: BucketTableCache::new(),
+            last_output: None,
+            output_changed: false,
+        }
+    }
+
+    pub async fn init(&mut self, init_epoch: EpochPair) -> StreamExecutorResult<()> {
+        // Init state tables.
+        self.count_state_table.init_epoch(init_epoch);
+        self.bucket_state_table.init_epoch(init_epoch);
+
+        // Refill row_count
+        let row_count_state = self.get_row_count_state().await?;
+        let row_count = Self::decode_row_count(&row_count_state)?;
+        self.row_count = row_count;
+        tracing::debug!(?row_count, "recovered row_count");
+
+        // Refill cache
+        self.refill_cache().await?;
+
+        // Update the last output downstream
+        let last_output = if row_count_state.is_none() {
+            None
+        } else {
+            Some(self.cache.get_output(row_count, self.quantile, self.base))
+        };
+        tracing::debug!(?last_output, "recovered last_output");
+        self.last_output = last_output;
+        Ok(())
+    }
+
+    async fn refill_cache(&mut self) -> StreamExecutorResult<()> {
+        let neg_bounds: (Bound<OwnedRow>, Bound<OwnedRow>) = (
+            Bound::Unbounded,
+            Bound::Excluded([Datum::from(ScalarImpl::Int16(0))].to_owned_row()),
+        );
+        let non_neg_bounds: (Bound<OwnedRow>, Bound<OwnedRow>) = (
+            Bound::Included([Datum::from(ScalarImpl::Int16(0))].to_owned_row()),
+            Bound::Unbounded,
+        );
+        #[for_await]
+        for keyed_row in self
+            .bucket_state_table
+            .rev_iter_with_prefix(&[Datum::None; 0], &neg_bounds, PrefetchOptions::default())
```

Comment:
May I ask why we use rev iter here? IMO the `self.cache.neg_buckets.insert` will work regardless of iter order?

Reply:
Fixed.